components/stratum/mining.c

```diff
@@ -133,27 +132,20 @@ double test_nonce_value(const bm_job *job, const uint32_t nonce, const uint32_t
 {
     double d64, s64, ds;
     unsigned char header[80];
-
-    // // TODO: use the midstate hash instead of hashing the whole header
-    // uint32_t rolled_version = job->version;
-    // for (int i = 0; i < midstate_index; i++) {
-    //     rolled_version = increment_bitmask(rolled_version, job->version_mask);
-    // }
-
-    // copy data from job to header
-    memcpy(header, &rolled_version, 4);
-    memcpy(header + 4, job->prev_block_hash, 32);
-    memcpy(header + 36, job->merkle_root, 32);
-    memcpy(header + 68, &job->ntime, 4);
-    memcpy(header + 72, &job->target, 4);
-    memcpy(header + 76, &nonce, 4);
-
     unsigned char hash_buffer[32];
     unsigned char hash_result[32];
 
-    // double hash the header
-    mbedtls_sha256(header, 80, hash_buffer, 0);
-    mbedtls_sha256(hash_buffer, 32, hash_result, 0);
+    // Use midstate instead of the whole hash
+    memcpy(hash_buffer, job->midstate, 32);  // use pre computed midstate
```

Comment:
the job midstate is calculated before the ASIC found a nonce, using the initial version (first field of the header) given by the stratum server. On some chip, hardware version rolling is performed, and the nonce come togheter with version_bits which should be added to the initial version from stratum server job. That's why we need to recompute the first midstate : to take into account the rolled version from asic hardware. So this improvement should be only used if the rolled_version in parameter of this function is equal to zero (chip didn't hardware rolled version).

Reply:
I see.
Changes:
Midstate Handling:
If the rolled_version is zero, the precomputed midstate (job->midstate) is used directly.
If the rolled_version is non-zero, the midstate is recalculated by incorporating the rolled_version in place of the initial version provided by the stratum server.
Midstate Calculation: When recalculating the midstate, the rolled_version is inserted at the start of the block header. This recalculated midstate is used in further hashing operations to ensure that the version bits from the ASIC are properly handled.
This ensures that the code efficiently uses the precomputed midstate when hardware version rolling is not required, and recalculates it only when necessary, improving overall mining efficiency.

`double test_nonce_value(const bm_job *job, const uint32_t nonce, const uint32_t rolled_version)
{
    double d64, s64, ds;
    unsigned char header[80];
    unsigned char hash_buffer[32];
    unsigned char hash_result[32];
    uint8_t midstate_data[64];

    // If rolled_version is 0, use precomputed midstate from the job
    if (rolled_version == 0)
    {
        memcpy(hash_buffer, job->midstate, 32);  // Use precomputed midstate
    }
    else
    {
        // Recompute midstate to account for rolled version
        memcpy(midstate_data, &rolled_version, 4);               // Copy rolled version
        memcpy(midstate_data + 4, job->prev_block_hash, 32);     // Copy previous block hash
        memcpy(midstate_data + 36, job->merkle_root, 28);        // Copy part of the merkle root

        // Compute the new midstate based on the rolled version
        midstate_sha256_bin(midstate_data, 64, hash_buffer);     // Calculate new midstate hash
        reverse_bytes(hash_buffer, 32);                          // Reverse midstate bytes to match format
    }

    // Complete the header with (ntime, target, nonce)
    memcpy(header, &job->ntime, 4);
    memcpy(header + 4, &job->target, 4);
    memcpy(header + 8, &nonce, 4);

    // Perform second round SHA-256 hashing
    mbedtls_sha256(header, 12, hash_buffer + 32, 0);  // Hash 12 bytes of remaining data
    mbedtls_sha256(hash_buffer, 64, hash_result, 0);  // Second round of SHA-256

    // Compute difficulty from the hash result
    d64 = truediffone;
    s64 = le256todouble(hash_result);
    ds = d64 / s64;

    return ds;
}
`